File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/number_of_islands/NumberofIslands3.java
package main.java.com.jiang.leetcode100.number_of_islands;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;

class Solution3 {

    class UnionFind {
        int count;
        int[] parent;
        int[] rank;

        public UnionFind(char[][] grid) {
            count = 0;
            int m = grid.length;
            int n = grid[0].length;
            parent = new int[m * n];
            rank = new int[m * n];
            for(int i = 0; i < m; i++) {
                for(int j = 0; j < n; j++) {
                    if(grid[i][j] == '1') {
                        parent[i * n + j] = i * n + j;
                        count++;
                    }
                    rank[i * n + j] = 0;
                }
            }
        }
        
        public int find(int i) {
            if(parent[i] != i) parent[i] = find(parent[i]);
            return parent[i];
        }

        public void union(int x, int y) {
            int rootx = find(x);
            int rooty = find(y);
            if(rootx != rooty) {
                if(rank[rootx] > rank[rooty]) {
                    parent[rooty] = rootx;
                } else if(rank[rootx] < rank[rooty]) {
                    parent[rootx] = rooty;
                } else {
                    parent[rooty] = rootx;
                    rank[rootx] += 1;
                }
                count--;
            }
        }

        public int getCount() {
            return count;
        }
    }

    public int numIslands(char[][] grid) {
        int m = grid.length;
        if(m == 0) return 0;
        int n = grid[0].length;

        UnionFind uf = new UnionFind(grid);
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1') {
                    grid[i][j] = '0';
                    if(i > 0 && grid[i - 1][j] == '1') {
                        uf.union(i * n + j, (i - 1) * n + j);
                    }
                    if(i < grid.length - 1 && grid[i + 1][j] == '1') {
                        uf.union(i * n + j, (i + 1) * n + j);
                    }
                    if(j > 0 && grid[i][j - 1] == '1') {
                        uf.union(i * n + j, i * n + j - 1);
                    }
                    if(j < grid[0].length - 1 && grid[i][j + 1] == '1') {
                        uf.union(i * n + j, i * n + j + 1);
                    }
                }
            }
        }

        
        return uf.getCount();
    }
}

public class NumberofIslands3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> inputs = new ArrayList<>();
        while(!scanner.hasNext("#")) {
            inputs.add(scanner.nextLine());
        }
        scanner.close();

        char[][] grid = new char[inputs.size()][inputs.get(0).length()];
        for(int i = 0; i < inputs.size(); i++) {
            for(int j = 0; j < inputs.get(i).length(); j++) {
                grid[i][j] = inputs.get(i).charAt(j);
            }
        }

        System.out.println(new Solution3().numIslands(grid));
    }
}

File: Leetcode100/leetcode100_python/NumberofIslands3.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        class UnionFind:

            def __init__(self):
                self.count = 0
                m, n = len(grid), len(grid[0])
                self.parent = [0] * (m * n)
                for i in range(m):
                    for j in range(n):
                        if grid[i][j] == '1':
                            self.parent[i * n + j] = i * n + j
                            self.count += 1

            def find(self, x):
                while self.parent[x] != x:
                    self.parent[x] = self.parent[self.parent[x]]
                    x = self.parent[x]
                return x

            def union(self, x, y):
                parentx = self.find(x)
                parenty = self.find(y)
                if parentx != parenty:
                    self.parent[parentx] = parenty
                    self.count -= 1
            
            def getCount(self):
                return self.count

        if not grid:
            return 0
        m = len(grid)
        n = len(grid[0])
        uf = UnionFind()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    grid[i][j] = '0'
                    if i > 0 and grid[i - 1][j] == '1':
                        uf.union(i * n + j, (i - 1) * n + j)
                    if i < len(grid) - 1 and grid[i + 1][j] == '1':
                        uf.union(i * n + j, (i + 1) * n + j)
                    if j > 0 and grid[i][j - 1] == '1':
                        uf.union(i * n + j, i * n + j - 1)
                    if j < len(grid[0]) - 1 and grid[i][j + 1] == '1':
                        uf.union(i * n + j, i * n + j + 1)
        return uf.getCount()


if __name__ == "__main__":
    n = int(input())
    grid = []
    for i in range(n):
        grid.append([num for num in input().split()])
    solution = Solution()
    print(solution.numIslands(grid))